### PR TITLE
Add `with_backlog` functionality to `TcpListener` and `UnixListener`

### LINKED ISSUE #94406: Add ability to specify backlog size for `TcpListener` and `UnixListener`
In socket-style network interfaces, the server side typically requests a "backlog" from the operating system. This controls how many requests the OS is willing to buffer before failing or ignoring new requests.

The backlogs used by `std::net::TcpListener::bind`, `std::os::unix::net::UnixListener::bind` and the nightly `std::os::unix::net::UnixListener::bind_addr` are currently hardwired to 128. This is a reasonable default, but there are situations where one might want to set a larger or smaller number. ~~Also, that 128 default should probably be a publicly-exposed constant.~~

To address these issues I have posted a PR that provides the following new APIs:

* ~~`std::net::TcpListener::DEFAULT_BACKLOG`~~
* `std::net::TcpListener::bind_with_backlog`
* ~~`std::os::unix::net::UnixListener::DEFAULT_BACKLOG`~~
* `std::os::unix::net::UnixListener::bind_with_backlog`
* `std::os::unix::net::UnixListener::bind_addr_with_backlog`

### PR DESCRIPTION
**Pull request by BartMassey**
Adds

* `std::net::TcpListener::DEFAULT_BACKLOG`
* `std::net::TcpListener::bind_with_backlog`
* `std::os::unix::net::UnixListener::DEFAULT_BACKLOG`
* `std::os::unix::net::UnixListener::bind_with_backlog`
* `std::os::unix::net::UnixListener::bind_addr_with_backlog`

Closes #94406

### CODE REVIEW

File: library/std/src/net/tcp.rs [hunk 1]
```diff
@@ -707,10 +707,53 @@ impl fmt::Debug for TcpStream {
...
+
     /// Creates a new `TcpListener` which will be bound to the specified
     /// address.
     ///
-    /// The returned listener is ready for accepting connections.
+    /// The given backlog specifies the maximum number of outstanding
+    /// connections that will be buffered in the OS waiting to be accepted by
+    /// [`TcpListener::accept`]. The backlog argument overrides the default
+    /// specified by [`TcpListener::DEFAULT_BACKLOG`]; that default is
+    /// reasonable for most use cases.
```
Comment (joshtriplett) [inline library/std/src/net/tcp.rs:723]:
```suggestion
    /// [`TcpListener::accept`]. The backlog argument overrides the default
    /// backlog used by [`TcpListener::bind`]; that default works reasonably
    /// for most use cases.
```

**Comment by BartMassey** [comment]
Have unpublished `DEFAULT_BACKLOG` for listeners and adjusted the docs accordingly.

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/94425) on wasm

**Comment by BartMassey** [comment]
@joshtriplett @Dylan-DPC Looks like I forgot a wasm stub somehow. Fixed.

**Comment by BartMassey** [comment]
Actually, looks like I forgot a bunch more stuff. Pls hold until I fix the rest of them.

**Comment by BartMassey** [comment]
What to do with platforms like sgx, sgx_fortanix and hermit that apparently support `bind()` but have no notion of backlog? I'm stuck for now.

**Comment by JohnCSimon** [comment]
@BartMassey 

Ping from triage: Can you please post the status of this PR? You can close it as inactive.
Thank you.

**Comment by BartMassey** [comment]
Sure, let's call it inactive for now. I'll try again once I've sorted some stuff.

**Pull request closed without merging**